MNT Apply ruff/flake8-implicit-str-concat rules (ISC)

DimitriPapadopoulos opened the pull request:
#### What does this implement/fix? Explain your changes.

Applying rule `ISC001` merges split string literals — usually as a result of black unfolding a line.

As for rule `ISC003`, it may appear it doesn't improve readability in some cases. However, I see such cases as black formatting issues.

#### Any other comments?

Not sure these rules should be enforced in CI, because I'm not certain they're always compatible with black.

[doc skip]

DimitriPapadopoulos committed:
MNT Apply ruff/flake8-implicit-str-concat rule ISC001

ISC001 Implicitly concatenated string literals on one line

DimitriPapadopoulos committed:
MNT Apply ruff/flake8-implicit-str-concat rule ISC003

ISC003 Explicitly concatenated string should be implicitly concatenated

adrinjalali commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -103,7 +103,7 @@ def plot_results(ax1, ax2, estimator, X, y, title, plot_title=False):
 estimators = [
     (
-        "Finite mixture with a Dirichlet distribution\nprior and " r"$\gamma_0=$",
+        "Finite mixture with a Dirichlet distribution\nprior and $\\gamma_0=$",
```
raw strings are easier to understand. Rather keep them than escaping characters.

DimitriPapadopoulos replied:
Sure I'll revert this, or perhaps change to split on `\n` which I find more readable:

```diff
-        "Finite mixture with a Dirichlet distribution\nprior and " r"$\gamma_0=$",
+        "Finite mixture with a Dirichlet distribution\n" r"prior and $\gamma_0=$",
```

adrinjalali commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -40,7 +40,7 @@
     for score_func in score_funcs:
         expected = (
-            r"Found input variables with inconsistent numbers " r"of samples: \[2, 3\]"
+            r"Found input variables with inconsistent numbers of samples: \[2, 3\]"
```
same here, raw string w/o escape characters is nicer.

DimitriPapadopoulos replied:
I did not remove the `r`, just merged two raw strings into a single raw string.

adrinjalali reviewed (commented):
Otherwise LGTM

adrinjalali reviewed (approved):
Please avoid force pushing since it makes review harder. It's easier to have the whole history in the PR.

DimitriPapadopoulos commented:
Sure, I could switch to `ruff format` in a different PR.

adrinjalali commented:
@Charlie-XIAO maybe you could have a look? Should be an easy review.

Charlie-XIAO commented on sklearn/tests/test_min_dependencies_readme.py:
```diff
@@ -33,9 +33,9 @@ def test_min_dependencies_readme():
-        + r"(([A-Za-z]+\-?)+)"
-        + r"(MinVersion\| replace::)"
-        + r"( [0-9]+\.[0-9]+(\.[0-9]+)?)"
+        r"(([A-Za-z]+\-?)+)"
```
```suggestion
        r"(([A-Za-z]+\-?)+)"
        r"(([A-Za-z]+(?:-[A-Za-z]+)*)?)"
```

Maybe? I'm not so familiar with regex myself though. CodeQL warned about exponential backtracking, so I'm removing the outer repetition `+`.

DimitriPapadopoulos replied:
I'm not sure about this one, it's unrelated to this PR. This is probably the result of a CodeQL upgrade.

Charlie-XIAO reviewed (commented):
This LGTM, thanks! I think we do need to fix CodeQL though.

DimitriPapadopoulos committed:
MNT Fix CodeQL

Inefficient regular expression

This part of the regular expression may cause exponential backtracking on strings starting with `.. |` and containing many repetitions of `A`.

Co-authored-by: Yao Xiao <108576690+Charlie-XIAO@users.noreply.github.com>

DimitriPapadopoulos commented on sklearn/tests/test_min_dependencies_readme.py:
```diff
@@ -33,9 +33,10 @@
-        + r"(([A-Za-z]+\-?)+)"
-        + r"(MinVersion\| replace::)"
-        + r"( [0-9]+\.[0-9]+(\.[0-9]+)?)"
+        r"(([A-Za-z]+\-?)+)"
```
CodeQL is still unhappy. I suggest leaving the regex as is for now. Either CodeQL runs on diffs (modified code) only, or this is the result of a CodeQL update. Either way, this is unrelated to this PR, or it the error will be fixed elsewhere any way.

Charlie-XIAO replied:
Okay, I'm fine with that.

DimitriPapadopoulos committed:
Revert "MNT Fix CodeQL"

This reverts commit 76b2c95f8a4de5d1db4b21d229b5b5a3d71c8901.
